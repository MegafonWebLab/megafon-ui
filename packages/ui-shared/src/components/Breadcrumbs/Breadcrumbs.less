@import '~@megafon/ui-core/styles/base.less';

@b: e('.mfui-breadcrumbs');

@{b} {
    display: flex;
    flex-wrap: wrap;

    &_color_default {
        @{b}__item:before {
            color: var(--content);
        }
    }

    &_color_black {
        @{b}__item:before {
            color: var(--stcBlack);
        }
    }

    &_color_white {
        @{b}__item:before {
            color: var(--stcWhite);
        }
    }

    &__icon {
        position: relative;

        width: 14px;
        height: 3px;
        margin-right: 10px;
    }

    &__item {
        position: relative;

        margin-right: 10px;
        padding-left: 22px;

        &:before {
            content: 'â†’';
            position: absolute;
            top: -2px;
            left: 0;
        }
    }

    &__item:first-child {
        padding-left: 0;

        &:before {
            display: none;
        }
    }

    &__item:last-child {
        pointer-events: none;
    }
}
