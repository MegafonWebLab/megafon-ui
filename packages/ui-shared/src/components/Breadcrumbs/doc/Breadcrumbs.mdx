---
name: Breadcrumbs
menu: Конструктор
route: components/breadcrumbs
---

import Example from './Breadcrumbs.example.mdx';
import Props from './Breadcrumbs.props.mdx';

# Breadcrumbs

<CopyToClipboardBox text="import { Breadcrumbs } from '@megafon/ui-shared';" />

<DoczTabs>
    <Example />
    <Props />
</DoczTabs>
