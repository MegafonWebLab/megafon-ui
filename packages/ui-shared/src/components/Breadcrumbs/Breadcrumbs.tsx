import React from 'react';
import { TextLink } from '@megafon/ui-core';
import { cnCreate, filterDataAttrs } from '@megafon/ui-helpers';
import PropTypes from 'prop-types';
import './Breadcrumbs.less';

type ItemObjectType = {
    title?: string;
    href?: string;
    /* @deprecated */
    component?: React.ReactElement;
};

type ItemType = JSX.Element | ItemObjectType;

export const TextColor = {
    DEFAULT: 'default',
    BLACK: 'black',
    WHITE: 'white',
} as const;

type TextColorType = typeof TextColor[keyof typeof TextColor];

export type Props = {
    dataAttrs?: {
        root?: Record<string, string>;
        link?: Record<string, string>;
    };
    className?: string;
    classes?: { item?: string };
    items: ItemType[];
    color?: TextColorType;
};

function isObjectElement(item: ItemType): item is ItemObjectType {
    return !React.isValidElement(item);
}

function isJSXElement(item: ItemType): item is JSX.Element {
    return React.isValidElement(item);
}

const cn = cnCreate('mfui-breadcrumbs');
const Breadcrumbs: React.FC<Props> = ({ items, color = 'default', className, classes = {}, dataAttrs }) => (
    <div {...filterDataAttrs(dataAttrs?.root)} className={cn({ color }, className)}>
        {items.map((item: ItemType, i: number): JSX.Element | null => {
            if (isJSXElement(item)) {
                return (
                    <div className={cn('item', classes.item)} key={i}>
                        {item}
                    </div>
                );
            }

            if (isObjectElement(item)) {
                return (
                    <div className={cn('item', classes.item)} key={item.component ? i : item.title}>
                        {item.component || (
                            <TextLink
                                href={item.href}
                                color={color}
                                dataAttrs={{
                                    root: dataAttrs?.link ? { ...filterDataAttrs(dataAttrs?.link, i + 1) } : undefined,
                                }}
                            >
                                {item.title}
                            </TextLink>
                        )}
                    </div>
                );
            }

            return null;
        })}
    </div>
);

Breadcrumbs.propTypes = {
    dataAttrs: PropTypes.shape({
        root: PropTypes.objectOf(PropTypes.string.isRequired),
        link: PropTypes.objectOf(PropTypes.string.isRequired),
    }),
    className: PropTypes.string,
    classes: PropTypes.shape({
        item: PropTypes.string,
    }),
    items: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.element.isRequired),
        PropTypes.arrayOf(
            PropTypes.shape({
                title: PropTypes.string,
                href: PropTypes.string,
                /* @deprecated */
                component: PropTypes.element,
            }).isRequired,
        ),
    ]).isRequired,
    color: PropTypes.oneOf(Object.values(TextColor)),
};

export default Breadcrumbs;
