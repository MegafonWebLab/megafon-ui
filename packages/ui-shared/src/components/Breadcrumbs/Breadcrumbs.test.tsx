import React from 'react';
import { render } from '@testing-library/react';
import Breadcrumbs, { Props } from './Breadcrumbs';

const items = [
    {
        title: 'МегаФон',
        href: '#',
    },
    {
        title: 'Мобильная связь',
        href: '#',
    },
    {
        title: 'Тарифы',
        href: '#',
    },
];

const dataAttrs: Props['dataAttrs'] = {
    root: {
        'data-testid': 'root',
    },
    item: {
        'data-testid': 'item',
    },
    itemInner: {
        'data-testid': 'itemInner',
    },
    itemTitle: {
        'data-testid': 'itemTitle',
    },
    itemMeta: {
        'data-testid': 'itemMeta',
    },
    link: {
        'data-testid': 'link',
    },
};

describe('<Breadcrumbs />', () => {
    it('should render Breadcrumbs', () => {
        const { container } = render(<Breadcrumbs items={items} />);

        expect(container).toMatchSnapshot();
    });

    it('should render Breadcrumbs with custom items', () => {
        const { container } = render(
            <Breadcrumbs
                items={[<a href="test1.com">Item 1</a>, <a href="test2.com">Item 2</a>, <span>Item 3</span>]}
            />,
        );

        expect(container).toMatchSnapshot();
    });

    it('should render with classes', () => {
        const { getByTestId } = render(
            <Breadcrumbs
                className="custom-class"
                classes={{
                    item: 'item-class',
                    itemTitle: 'item-title-class',
                    lastItemTitle: 'last-item-title-class',
                    link: 'link-class',
                }}
                items={items}
                dataAttrs={dataAttrs}
            />,
        );

        expect(getByTestId('root')).toHaveClass('custom-class');
        expect(getByTestId('item[1]')).toHaveClass('item-class');
        expect(getByTestId('itemTitle[1]')).toHaveClass('item-title-class');
        expect(getByTestId('itemTitle[3]')).toHaveClass('last-item-title-class');
        expect(getByTestId('link[1]')).toHaveClass('link-class');
    });

    it('should render with color', () => {
        const { getByTestId } = render(<Breadcrumbs color="white" items={items} dataAttrs={dataAttrs} />);

        expect(getByTestId('root')).toHaveClass('mfui-breadcrumbs_color_white');
        expect(getByTestId('link[1]')).toHaveClass('mfui-text-link_color_white');
        expect(getByTestId('itemInner[3]')).toHaveClass('mfui-breadcrumbs-item_color_white');
    });

    it('should render when hasMicrodata is true', () => {
        const { getByTestId } = render(<Breadcrumbs hasMicrodata items={items} dataAttrs={dataAttrs} />);
        const root = getByTestId('root');
        const itemInner = getByTestId('itemInner[1]');

        expect(root).toHaveAttribute('itemscope');
        expect(root).toHaveAttribute('itemtype', 'https://schema.org/BreadcrumbList');
        expect(itemInner).toHaveAttribute('itemscope');
        expect(itemInner).toHaveAttribute('itemprop', 'itemListElement');
        expect(itemInner).toHaveAttribute('itemtype', 'https://schema.org/ListItem');
        expect(getByTestId('link[1]')).toHaveAttribute('itemprop', 'item');
        expect(getByTestId('itemTitle[1]')).toHaveAttribute('itemprop', 'name');
        expect(getByTestId('itemMeta[1]')).toHaveAttribute('itemprop', 'position');
    });

    it('should render when hasMicrodata is false', () => {
        const { getByTestId, queryByTestId } = render(
            <Breadcrumbs hasMicrodata={false} items={items} dataAttrs={dataAttrs} />,
        );
        const root = getByTestId('root');
        const itemInner = getByTestId('itemInner[1]');

        expect(root).not.toHaveAttribute('itemscope');
        expect(root).not.toHaveAttribute('itemtype');
        expect(itemInner).not.toHaveAttribute('itemscope');
        expect(itemInner).not.toHaveAttribute('itemprop');
        expect(itemInner).not.toHaveAttribute('itemtype');
        expect(getByTestId('link[1]')).not.toHaveAttribute('itemprop');
        expect(getByTestId('itemTitle[1]')).not.toHaveAttribute('itemprop');
        expect(queryByTestId('itemMeta[1]')).not.toBeInTheDocument();
    });
});
