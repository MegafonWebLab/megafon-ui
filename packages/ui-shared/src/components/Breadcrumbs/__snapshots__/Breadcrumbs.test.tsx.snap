// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs should render component 1`] = `
<div
  className="mfui-breadcrumbs mfui-breadcrumbs_color_default"
>
  <div
    className="mfui-breadcrumbs__item"
    key="МегаФон"
  >
    <TextLink
      color="default"
      dataAttrs={
        Object {
          "root": undefined,
        }
      }
      href="#"
    >
      МегаФон
    </TextLink>
  </div>
  <div
    className="mfui-breadcrumbs__item"
    key="Мобильная связь"
  >
    <TextLink
      color="default"
      dataAttrs={
        Object {
          "root": undefined,
        }
      }
      href="#"
    >
      Мобильная связь
    </TextLink>
  </div>
  <div
    className="mfui-breadcrumbs__item"
    key="Тарифы"
  >
    <TextLink
      color="default"
      dataAttrs={
        Object {
          "root": undefined,
        }
      }
      href="#"
    >
      Тарифы
    </TextLink>
  </div>
</div>
`;

exports[`Breadcrumbs should render with optional props 1`] = `
<div
  className="mfui-breadcrumbs mfui-breadcrumbs_color_white custom-class-name"
  data-testid="root-test"
>
  <div
    className="mfui-breadcrumbs__item item-custom-class-name"
    key="МегаФон"
  >
    <TextLink
      color="white"
      dataAttrs={
        Object {
          "root": Object {
            "data-testid": "link-test[1]",
          },
        }
      }
      href="#"
    >
      МегаФон
    </TextLink>
  </div>
  <div
    className="mfui-breadcrumbs__item item-custom-class-name"
    key="Мобильная связь"
  >
    <TextLink
      color="white"
      dataAttrs={
        Object {
          "root": Object {
            "data-testid": "link-test[2]",
          },
        }
      }
      href="#"
    >
      Мобильная связь
    </TextLink>
  </div>
  <div
    className="mfui-breadcrumbs__item item-custom-class-name"
    key="Тарифы"
  >
    <TextLink
      color="white"
      dataAttrs={
        Object {
          "root": Object {
            "data-testid": "link-test[3]",
          },
        }
      }
      href="#"
    >
      Тарифы
    </TextLink>
  </div>
</div>
`;
